Mod7.1: Beyond MVC; Stubs & Mocks

ppj opened the pull request:
Decorator for Video rating display
Service for User sign-up handling

ppj committed:
Decorator for video to show rating string

ppj committed:
Create a service object for handling the user sign-up

Rearrange the specs according to the concerns

ppj commented on app/controllers/users_controller.rb:
```diff
@@ -17,25 +17,14 @@ def new_invited
-      end
+    result = UserSignup.perform(user: @user,
+                                invitation_token: params[:invitation_token],
+                                stripe_token: params[:stripeToken])
```
I like this way of delegating better than `UserSignup.new(blah blah).perform`. We try to use [this pattern at work](http://webuild.envato.com/blog/a-case-for-use-cases/). Thoughts?

tomtomecek replied:
This is good too :) The pattern before `Service.new(...).perform` can be potentially more clear on whats going on, - might be better for beginners.

ppj replied:
Agreed @tomtomecek. May be because I am used to seeing it now is why I find it easier to read. I did take me a couple of reads to get my head around it :smile:

tomtomecek commented on app/services/user_signup.rb:
```diff
@@ -0,0 +1,52 @@
+    new(user, invitation_token, stripe_token).perform
+  end
+
+  def initialize(user, invitation_token, stripe_token)
```
I would certainly recommend setting `invitation_token` to `nil` as default. Not all sign ups will be through invitation. Also its better when you build tests.

tomtomecek commented on app/services/user_signup.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def perform
+    if user.valid?
+      Stripe.api_key = ENV["STRIPE_SECRET_KEY"]
```
I would move this to initializer.

tomtomecek commented on app/services/user_signup.rb:
```diff
@@ -0,0 +1,52 @@
+    else
+       @error_message = "Please fix the highlighted errors before continuing..."
+    end
+    self
```
Good job :)

tomtomecek commented on spec/services/user_signup_spec.rb:
```diff
@@ -0,0 +1,116 @@
+        after { ActionMailer::Base.deliveries.clear }
+
+        it "creates new user" do
+          perform
```
On 1 side this looks nice, - not many details. On the other side, your specs are documenting the code and refactoring some details above will make it more unclear itself. So, imagine you have bug or error at certain lines. You will follow to those lines and then you will have to most likely check out the lines that are way above.

When you have before blocks and lots of context, reading and understanding 1 spec will be way more difficult. Refactoring and drying specs out is not always a good option to do.

ppj replied:
That's right. I have added a `subject` to each `context` and gotten rid of some unnecessary `let`s.

tomtomecek commented on spec/services/user_signup_spec.rb:
```diff
@@ -0,0 +1,116 @@
+
+        it "does not create a new user" do
+          perform
+          expect(User.count).to eq 0
```
I think you can refactor this to

``` ruby
expect { perform }.to_not change { User.count }
```

ppj committed:
Address review comments

- default invitation_token should be nil
- set the Stripe.api_key in the constructor
- undry the specs to serve as better documentation

ppj committed:
Extract some private methods to abstract away stuff & for readability
